packages/storage-plus/src/prefix.rs

```diff
@@ -124,6 +129,96 @@ where
     }
 }
 
+#[derive(Clone)]
+pub struct Prefix2<K, T>
+where
+    K: Deserializable,
+    T: Serialize + DeserializeOwned,
+{
+    /// all namespaces prefixes and concatenated with the key
+    storage_prefix: Vec<u8>,
+    // see https://doc.rust-lang.org/std/marker/struct.PhantomData.html#unused-type-parameters for why this is needed
+    data: PhantomData<T>,
+    pk_name: Vec<u8>,
+    de_fn: DeserializeKvFn<K, T>,
+}
+
+impl<K, T> Deref for Prefix2<K, T>
+where
+    K: Deserializable,
+    T: Serialize + DeserializeOwned,
+{
+    type Target = [u8];
+
+    fn deref(&self) -> &[u8] {
+        &self.storage_prefix
+    }
+}
+
+impl<K, T> Prefix2<K, T>
+where
+    K: Deserializable,
+    T: Serialize + DeserializeOwned,
+{
+    pub fn new(top_name: &[u8], sub_names: &[&[u8]]) -> Self {
+        Prefix2::with_deserialization_function(top_name, sub_names, &[], |_, _, kv| {
+            deserialize_kv::<K, T>(kv)
+        })
+    }
+
+    pub fn with_deserialization_function(
+        top_name: &[u8],
+        sub_names: &[&[u8]],
+        pk_name: &[u8],
+        de_fn: DeserializeKvFn<K, T>,
+    ) -> Self {
+        let storage_prefix = nested_namespaces_with_key(&[top_name], sub_names, b"");
+        Prefix2 {
+            storage_prefix,
+            data: PhantomData,
+            pk_name: pk_name.to_vec(),
+            de_fn,
+        }
+    }
+
+    pub fn range_de<'a>(
+        &self,
+        store: &'a dyn Storage,
+        min: Option<Bound>,
+        max: Option<Bound>,
+        order: Order,
+    ) -> Box<dyn Iterator<Item = StdResult<Pair2<K::Output, T>>> + 'a>
+    where
+        T: 'a,
+        K::Output: 'a,
+    {
+        let de_fn = self.de_fn;
+        let pk_name = self.pk_name.clone();
+        let mapped = range_with_prefix(store, &self.storage_prefix, min, max, order)
+            .map(move |kv| (de_fn)(store, &*pk_name, kv));
+        Box::new(mapped)
+    }
+
+    pub fn keys_de<'a>(
+        &self,
+        store: &'a dyn Storage,
+        min: Option<Bound>,
+        max: Option<Bound>,
+        order: Order,
+    ) -> Box<dyn Iterator<Item = StdResult<K::Output>> + 'a>
+    where
+        T: 'a,
+        K::Output: 'a,
```

Comment:
I wonder about this guard. Can't we just define `K::Output: 'static`, since we only deserialise into owned objects?

I mean this is true, `'static: 'a`, but it would be more clear. I guess with both K and T. If this is non-trivial to change, we can look at that in a different PR.

Reply:
Good idea. I was just following the compiler's guidelines here.